trajectory_utils/src/trajectory_utils/trajectory_utils.cpp

```diff
@@ -0,0 +1,175 @@
+  {  // Check for curvature of 0.
+    return std::numeric_limits<double>::infinity();
+  }
+  return std::sqrt(fabs(lateral_accel_limit / curvature));
+}
+
+std::vector<double> constrained_speeds_for_curvatures(std::vector<double> curvatures, double lateral_accel_limit)
+{
+  std::vector<double> out;
+  for (double k : curvatures)
+  {
+    out.push_back(trajectory_utils::constrain_speed_for_curvature(k, lateral_accel_limit));
+  }
+
+  return out;
+}
+
+std::vector<double> apply_accel_limits_by_distance(std::vector<double> downtracks, std::vector<double> speeds,
+                                                   double accel_limit, double decel_limit)
+{
+  if (downtracks.size() != speeds.size())
+  {
+    throw std::invalid_argument("Downtracks and speeds do not have the same size");
+  }
+
+  if (accel_limit <= 0 || decel_limit <= 0)
+  {
+    throw std::invalid_argument("Accel and Decel limits should be positive");
+  }
+
+  std::vector<double> output;
+  output.reserve(downtracks.size());
+
+  if (downtracks.size() == 0)
+  {
+    return output;
+  }
+
+  output.push_back(speeds[0]);  // First point will be unchanged
+
+  for (size_t i = 1; i < downtracks.size(); i++)
+  {
+    double delta_d = downtracks[i] - downtracks[i - 1];
+    double prev_speed = output.back();
+    double cur_speed = speeds[i];
+    double new_speed = cur_speed;
+    if (cur_speed > prev_speed)
+    {  // Acceleration case
+      new_speed = std::min(cur_speed, sqrt(prev_speed * prev_speed + 2 * accel_limit * delta_d));
+    }
+    else if (cur_speed < prev_speed)
+    {  // Deceleration case
+      new_speed = std::max(cur_speed, sqrt(prev_speed * prev_speed - 2 * decel_limit * delta_d));
+    }
+
+    new_speed = std::max(0.0, new_speed);
+
+    output.push_back(new_speed);
+  }
+
+  return output;
+}
+
+std::vector<double> apply_accel_limits_by_time(std::vector<double> times, std::vector<double> speeds,
+                                               double accel_limit, double decel_limit)
+{
+  if (times.size() != speeds.size())
+  {
+    throw std::invalid_argument("Times and speeds do not have the same size");
+  }
+
+  if (accel_limit <= 0 || decel_limit <= 0)
+  {
+    throw std::invalid_argument("Accel and Decel limits should be positive");
+  }
+
+  std::vector<double> output;
+  output.reserve(times.size());
+
+  if (times.size() == 0)
+  {
+    return output;
+  }
+
+  output.push_back(speeds[0]);  // First point will be unchanged
+
+  for (size_t i = 1; i < times.size(); i++)
+  {
+    double delta_t = times[i] - times[i - 1];
+    double prev_speed = output.back();
+    double cur_speed = speeds[i];
+    double new_speed = cur_speed;
+    if (cur_speed > prev_speed)
+    {  // Acceleration case
+      new_speed = std::min(cur_speed, prev_speed + accel_limit * delta_t);
+    }
+    else if (cur_speed < prev_speed)
+    {  // Deceleration case
+      new_speed = std::max(cur_speed, prev_speed - decel_limit * delta_t);
+    }
+    else
```

Comment:
Does not matter much but I suggest remove  "else" to be consistent with the previous function.